### PR TITLE
Anisha/backend flow

### PR DESCRIPTION
**Pull request by anishalatchman**
# Description

- Created clean architecture folder in frontend (DAO, Interactor, Controller)
- Made recover session page functional: recovered flow in DB from sessionID
- Replaced navState with SessionID. New SessionContext format is [sessionID, setSessionID, transcriptID, setTranscriptID]
- WORKING EVERYTHING
- Transcript scroller works
- Recover session works
- Every functionality you can imagine is functional :))
- Please test in every way shape or form to find edge cases or break things

NOTE: Fixed visual glitch of buttons misaligning a swell

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [x] Vanquishing our React and asynchronous enemies with the power of stackoverflow and await

# How Has This Been Tested?

Test suite to be added once flow parameters have been updated.

Console.log tests currently ensure:
- [X] Flow object is retrived from axios call in recoverSessionDAO
- [X] First question is accessed from flow object in recoverSessionInteractor
- [x] Manual testing has been done by Jacob :)). More testing needed from everyone

# Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: mern/client/src/Components/Navbar/navbar.component.js [hunk 1]
```diff
@@ -28,17 +26,17 @@ export default function Navbar() {
...
-          SESSION ID: {sessionid}  
+        <h2 className="font-nunito font-medium justify-center">
+          SESSION ID: {sessionID}  
         </h2>
         <GenericButton
         buttonType="nav"
         onClick={() => null}
         disabled={false}
-        text={"SAVE"}
+        text={"SAVE SESSION"}
```
Comment (EmmaHkw) [inline mern/client/src/Components/Navbar/navbar.component.js:39]: Oh this is much clearer good job!

Comment (EmmaHkw) [inline mern/client/src/Components/Navbar/navbar.component.js:39]: this is much clearer good job!

File: mern/client/src/DAO/recoverSessionDAO.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// Returns flow json object with matching sessionID
+export async function getFlowBySessionID(sessionID) {
+    try {
+      const response = await axios.get(`http://localhost:5000/flow/${sessionID}`, sessionID);
+      const { data } = response;
+      // return the flow object
+      return data;
+
+    } catch (e) {
+      return false;
```
Comment (EmmaHkw) [inline mern/client/src/DAO/recoverSessionDAO.js:13]: would this return anything to the user or is this return only for the function?

File: mern/client/src/screens/UploadTranscript/uploadTranscript.js [hunk 3]
```diff
@@ -128,8 +127,10 @@ function UploadTranscript() {
               flowUploader(flowName.name, files).then((response) => {
                 if (response) {
                   alert("Your flow name has been set to: " + flowName.name);
+                  // console.log(response)
+                  // FIX SESSIONID HERE!!!
```
Comment (EmmaHkw) [inline mern/client/src/screens/UploadTranscript/uploadTranscript.js:131]: Great job!

File: mern/client/src/DAO/recoverSessionDAO.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import axios from 'axios';
+// DAO is the inner most layer because DB stores our entities (flow, QA, transcript)
+
+// Returns flow json object with matching sessionID
+export async function getFlowBySessionID(sessionID) {
+    try {
+      const response = await axios.get(`http://localhost:5000/flow/${sessionID}`, sessionID);
```
Comment (EmmaHkw) [inline mern/client/src/DAO/recoverSessionDAO.js:7]: Good job in using clean architecture to have the backend data call in the DAO!

File: mern/client/src/screens/RecoverSession/recoverSession.js [hunk 5]
```diff
@@ -6,23 +6,24 @@ import { useNavigate } from "react-router-dom";
...
 
   const handleSubmit = (event) => {
+    // prevent page reload
     event.preventDefault();
     // show navbar buttons for starting-intent page
-    setNavState(true);
-    alert(`The session ID you entered was: ${sessionid}`);
-  };
+    setSessionID(inputText);
+    PageChange("/startingintent")
```
Comment (EmmaHkw) [inline mern/client/src/screens/RecoverSession/recoverSession.js:21]: I am just wondering, would this take the users to the first intents based on the transcript or the point they saved last time?

File: mern/client/src/screens/StartingIntent/StartingIntent.js [hunk 6]
```diff
@@ -1,32 +1,45 @@
...
   const [intentState] = useContext(IntentContext);
-  const Navigate = useNavigate();
-  const PageChange = (url) => {
-    Navigate(url);
-  };
+  // const Navigate = useNavigate();
+
+  // const PageChange = (url) => {
+  //   Navigate(url);
+  // };
```
Comment (EmmaHkw) [inline mern/client/src/screens/StartingIntent/StartingIntent.js:23]: maybe you can delete this and the commented lines above if it is not useful anymore? totally up to you though! :))

File: mern/server/controller/flow.js [hunk 7]
```diff
@@ -1,3 +1,6 @@
+// DATA ACCESS OBJECT - This file makes get calls to the MongoDB,
```
Comment (EmmaHkw) [inline mern/server/controller/flow.js:1]: Just wondering if there is a reason the DAO is in the controller's folder?😅

File: mern/server/controller/question_answer.js [hunk 8]
```diff
@@ -1,3 +1,6 @@
+// DATA ACCESS OBJECT - This file makes get calls to the MongoDB,
```
Comment (EmmaHkw) [inline mern/server/controller/question_answer.js:1]: same as above 😅

File: mern/server/controller/transcript.js [hunk 9]
```diff
@@ -1,3 +1,6 @@
+// DATA ACCESS OBJECT - This file makes get calls to the MongoDB,
```
Comment (EmmaHkw) [inline mern/server/controller/transcript.js:1]: and here😅😅

File: mern/client/src/Components/Navbar/navbar.component.js [hunk 10]
```diff
@@ -7,17 +7,15 @@ import { SessionContext } from "../../Contexts/sessionProvider";
...
-  // };
+  const [sessionID, setSessionID, , ] = useContext(SessionContext)
 
   return (
     <nav className="navbar-bg">
-      <Link to="/" className="nav-link" onClick={() => {setNavState(false)}}>
+      <Link to="/" className="nav-link" onClick={() => {
+        setSessionID(false)
+        // reset sessionID on return to home page
+        setSessionID()
```
Comment (EmmaHkw) [inline mern/client/src/Components/Navbar/navbar.component.js:17]: hmm wait why do you set it as false and then as empty?

**Review by EmmaHkw — CHANGES_REQUESTED** [review]
Hey Anisha! Great work thank you, especially with the clean architecture! However, regarding that, I am wondering why the backend DAOs are in the controller folder? I also added additional comments throughout! :)

File: mern/client/src/Components/Navbar/navbar.component.js (hunk 10 above)
Comment (anishalatchman) [reply mern/client/src/Components/Navbar/navbar.component.js:16]: Ah thank you for the catch! The set false was from when we had navState, but that's outdated now!

File: mern/server/controller/flow.js (hunk 7 above)
Comment (anishalatchman) [reply mern/server/controller/flow.js]: This has been resolved in Terry's recent clean architecture branch!

File: mern/client/src/screens/StartingIntent/StartingIntent.js (hunk 6 above)
Comment (anishalatchman) [reply mern/client/src/screens/StartingIntent/StartingIntent.js:23]: Deleted it! I thought we might need it in the future, but I'm now realizing the starting-intent page doesn't reroute to any other pages.

File: mern/client/src/DAO/recoverSessionDAO.js (hunk 2 above)
Comment (anishalatchman) [reply mern/client/src/DAO/recoverSessionDAO.js:13]: This returns the flow object from the DB up to the interactor, then the controller, where the QA's are extracted and displayed on the screen for the user.

Comment (EmmaHkw) [reply mern/client/src/DAO/recoverSessionDAO.js:13]: I see thank you!

**Review by EmmaHkw — DISMISSED** [review]

File: mern/client/src/Interactor/recoverSessionInteractor.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+export async function sendToDAO(sessionID) {
+    // returns a flow promise object
+    const flow = await getFlowBySessionID(sessionID);
+    console.log("this is the flow", flow);
+    return flow;
+
+}
+
+// returns first QA pair from given flow promise object
+export function getQAFromFlow(flowPromise) {
```
Comment (TerryTian21) [inline mern/client/src/Interactor/recoverSessionInteractor.js:15]: What is the purpose of this function, I don't see why you would ever need to fetch only the "id" for the starting 5 questions.

File: mern/client/src/Controller/recoverSessionController.js [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import { sendToDAO, getQAFromFlow } from "../Interactor/recoverSessionInteractor"
+
+// returns starting QA pair from session ID
+export async function recoverFlow(sessionID) {
+    // returns flow promise object
+    const flow = await sendToDAO(sessionID);
```
Comment (TerryTian21) [inline mern/client/src/Controller/recoverSessionController.js:6]: This part seems a little weird. You are getting flow as a return, which is a promise here. Then you are sending flow into the getQAFromFlow function. This means you retrieved a response from the DAO, then sent it back to the DAO. 

I get that the goal is to get the first starting question, but realistically these shows be separated into different controllers and interactors. 

The way that you want to do it is, to pass the flow response all the way back up to where you need to use it. 
Then, access the first element of the questions list at the top level which gives you the id of the first question.
After, you go through a QA controller which then calls the QA interactor which then calls the QA DAO that fetches the QA object and returns it to the top layer.

File: mern/client/src/Interactor/recoverSessionInteractor.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { getFlowBySessionID } from "../DAO/recoverSessionDAO"
+//  import dao to call axios function
+// manipulate data
+
+// sessionID does not need to be manipulated so this function sends it straight to DAO
+export async function sendToDAO(sessionID) {
+    // returns a flow promise object
+    const flow = await getFlowBySessionID(sessionID);
+    console.log("this is the flow", flow);
+    return flow;
```
Comment (TerryTian21) [inline mern/client/src/Interactor/recoverSessionInteractor.js:10]: Keep in mind that the flow object is a promise, so you would need to access the contents with a .then

**Review by TerryTian21 — COMMENTED** [review]
Most of the front-end changes look fine to me.

Your recover session interactor file seems a little bit weird. But I think it is fine if you were to argue that you create the interactor based on use case rather than based on entity.

File: mern/client/src/Interactor/recoverSessionInteractor.js (hunk 11 above)
Comment (anishalatchman) [reply mern/client/src/Interactor/recoverSessionInteractor.js:15]: This function is meant to return the current_question QA object from the flow, to display on the starting-intent screen. I will update it now that the flow's been updated.

File: mern/client/src/Controller/recoverSessionController.js (hunk 12 above)
Comment (anishalatchman) [reply mern/client/src/Controller/recoverSessionController.js:6]: Oops, looks like the comments were out of date here. Thanks for the flag. The flow return is a json object, not a promise.

I see what you are saying about retrieving the QA object through a seperate controller/interactor/dao. I mistakenly thought the current_question parameter in the flow was returning the QA object, not just the ID. Will make the fix to retrieve the QA object itself!

File: mern/client/src/Interactor/recoverSessionInteractor.js (hunk 13 above)
Comment (anishalatchman) [reply mern/client/src/Interactor/recoverSessionInteractor.js:10]: *flow object is a json object, not a promise. Comments have been updated to reflect that.

File: mern/client/src/Contexts/flowProvider.js [hunk 14]
```diff
@@ -3,13 +3,14 @@ import React from "react";
 export const FlowContext = React.createContext();
```
Comment (TerryTian21) [inline mern/client/src/Contexts/flowProvider.js:3]: Nice good to see the documentation for our contexts.

File: mern/client/src/DAO/flowDAO.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+import axios from 'axios';
+// DAO is the inner most layer because DB stores our entities (flow, QA, transcript)
+
+// Returns flow json object with matching sessionID
+export async function getFlowBySessionID(sessionID) {
+    try {
+      const response = await axios.get(`http://localhost:5000/flow/${sessionID}`, sessionID);
```
Comment (TerryTian21) [inline mern/client/src/DAO/flowDAO.js:7]: change localhost to process.env

**Review by TerryTian21 — APPROVED** [review]
The DAO, Controller, and Interactor folders in SRC seem to not be used as there are repeated folders with the same name and functions in utils. There can be deleted.
